File: BOJ/Gold IV/[1967] 트리의 지름/diameterOfTree.py
import sys
sys.setrecursionlimit(10 ** 6)

def lineLength(graph, node, visited):
    # 현재 노드 방문 처리
    visited[node[0]] = True
    length = 0
    lastNode = 0
    # 깊이 우선 탐색으로 가중치 길이 구하기
    for i in graph[node[0]]:
        if not visited[i[0]]:
            # 하나의 일직선 길이 구하기
            tmpLength, tmpNode = lineLength(graph, i, visited)
            tmpLength += i[1]
            # 현재 구해진 길이보다 길다면 tmp로 값 변경
            if tmpLength > length:
                length = tmpLength
                lastNode = tmpNode

    # 마지막 노드일 경우 자신의 노드 반환
    if length == 0:
        lastNode = node[0]
    return length, lastNode


n = int(input())
# n이 1이면 0 반환
if n == 1:
    print(0)
else:
    # graph 만들기
    graph = [[] for _ in range(n + 1)]
    for i in range(1, n):
        pNode, cNode, weight = map(int, input().split())
        # 부모노드에 자식노드, 가중치 저장
        graph[pNode].append((cNode, weight))
        # 자식노드에 부모노드, 가중치 저장
        graph[cNode].append((pNode, weight))

    # leaf 노드 구하기
    leafNode = 0
    for i in range(n):
        if len(graph[i + 1]) == 1:
            leafNode = i + 1

    answer = 0

    # 첫 leaf 노드에서 가장 긴 경로의 leaf Node 구하기
    # 해당 leaf 노드에서 가장 긴 경로 구하기
    for _ in range(2):
        # 방문 처리 리스트
        visited = [False] * (n + 1)
        # 시작 노드 방문 처리
        visited[leafNode] = True

        # 시작 노드에 연결된 노드 방향으로 길이 구하기
        length = graph[leafNode][0][1]
        tmpLength, leafNode = lineLength(graph, graph[leafNode][0], visited)
        # 이전 길이보다 길 경우 answer 값 변경
        answer = length + tmpLength
    print(answer)

File: BOJ/Gold IV/[1967] 트리의 지름/diameterOfTree.kt
class Tree(var n: Int) {
    private var graph: ArrayList<ArrayList<ArrayList<Int>>> = ArrayList<ArrayList<ArrayList<Int>>>()
    private var visited: ArrayList<Boolean> = ArrayList<Boolean>()
    private var leafNode: ArrayList<Int> = ArrayList<Int>()

    init {
        for(i in 0..this.n) {
            this.graph.add(ArrayList<ArrayList<Int>>())
            this.visited.add(false)
        }
        makeTree()
    }

    //  트리 만들기
    private fun makeTree() {
        for(i in 1 until this.n) {
            var nodeWeight = readLine()!!.split(" ").map{ it.toInt() }
            // 부모노드에 (자식노드, 가중치) 저장
            var edge = arrayListOf(nodeWeight[1], nodeWeight[2])
            this.graph[nodeWeight[0]].add(edge)
            // 자식노드에 (부모노드, 가중치) 저장
            edge = arrayListOf(nodeWeight[0], nodeWeight[2])
            this.graph[nodeWeight[1]].add(edge)
        }
        
        // leafNode 만들기
        for(i in 1..this.n) {
            if(this.graph[i].size == 1) {
                this.leafNode.add(i)
            }
        }
    }

    // 노드의 가장 먼 노드까지의 길이 구하기
    private fun lineLength(node: ArrayList<Int>): Int {
        // 현재 노드 방문 처리
        this.visited[node[0]] = true
        var length = 0
        // 깊이 우선 탐색
        for (i in this.graph[node[0]]) {
            if(!this.visited[i[0]]) {
                var tmp = i[1] + lineLength(i)
                // 현재 구해진 길이보다 길다면 tmp 값으로 변경
                length = if(length < tmp) tmp else length
            }
        }
        return length
    }

    // 결과 구하기
    fun getAnswer() {
        var answer = 0
        for(idx in this.leafNode) {
            // 방문 처리 리스트
            for(i in 0..this.n) {
                this.visited[i] = false
            }
            // 시작노드 방문 처리
            this.visited[idx] = true

            // 시작노드에 연결된 노드 방향으로 길이 구하기
            var length = graph[idx][0][1] + lineLength(graph[idx][0])
            // 이전 길이보다 길경우 answer 값 변경
            answer = if(answer < length) length else answer
        }
        println(answer)
    }

}

fun main() {
    var n: Int = readLine()!!.toInt()
    var tree = Tree(n)
    tree.getAnswer()
}
